polus-label-to-vector-plugin/src/main.py

```diff
@@ -1,190 +1,378 @@
+from concurrent.futures import wait
 from multiprocessing import cpu_count
 from pathlib import Path
+
 import filepattern
-import numpy as np
-import zarr
-from bfio.bfio import BioReader, BioWriter
+import numpy
 import torch
-
+import zarr
+from bfio.bfio import BioReader
+from bfio.bfio import BioWriter
 from cellpose import dynamics
 
 logging.getLogger('cellpose').setLevel(logging.CRITICAL)
 
 TILE_SIZE = 2048
 TILE_OVERLAP = 512
 
-def flow_thread(input_path: Path,
-                zfile: Path,
-                use_gpu: bool,
-                dev: torch.device,
-                x: int,
-                y: int,
-                z: int) -> bool:
-    """ Converts labels to flows
+
+def clean_3d_flow(mask_slice, use_gpu, device) -> numpy.ndarray:
+    """ Adjusts the labels in the given slice before using cellpose to calculate flows.
+
+    Args:
+        mask_slice: A 2d slice from a 3d image.
+        use_gpu: Whether to use the GPU
+        device: Which CUDA device to sue
+
+    Returns:
+        2d-flow for the given slice. This flow is aggregated by the calling function.
+    """
+    function = dynamics.masks_to_flows_gpu if use_gpu else dynamics.masks_to_flows_cpu
+
+    if numpy.any(mask_slice):
+        shape = numpy.shape(mask_slice)
+        labels, mask_slice = numpy.unique(mask_slice, return_inverse=True)
+        mask_slice = mask_slice.reshape(shape)
+        if len(labels) == 1:
+            mask_slice += 1
+        flow_subset = function(mask_slice, device=device)[0]
+    else:
+        flow_subset = numpy.zeros_like(mask_slice, dtype=numpy.float32)
+
+    return flow_subset
+
+
+def flow_thread_3d(
+        input_path: Path,
+        zarr_path: Path,
+        use_gpu: bool,
+        device: torch.device,
+        coordinates: tuple[int, int],
+) -> bool:
+    """ Converts labels to flows.
 
     This function converts labels in each tile to vector field.
 
     Args:
-        input_path(path): Path of input image collection
-        zfile(path): Path where output zarr file will be saved
-        x(int): Start index of the tile in x dimension of image
-        y(int): Start index of the tile in y dimension of image
-        z(int): Z slice of the  image
+        input_path: Path of input image collection.
+        zarr_path: Path where output zarr file will be saved.
+        use_gpu: Path where output zarr file will be saved.
+        device: Path where output zarr file will be saved.
+        coordinates: (x, y) coordinates of the tile in the image.
 
     """
-    
+
     logging.basicConfig(format='%(asctime)s - %(name)-8s - %(levelname)-8s - %(message)s',
                         datefmt='%d-%b-%y %H:%M:%S')
-    logger = logging.getLogger("flow")
-    logger.setLevel(logging.INFO)
+    flow_logger = logging.getLogger("flow_3d")
+    flow_logger.setLevel(logging.DEBUG)  # TODO: Change back to INFO
+
+    # noinspection PyTypeChecker
+    root = zarr.open(str(zarr_path))[0]
+    x, y = coordinates
```

Comment:
For 3d, should `coordinates` be a 3-tuple? Right now, I see you take all z-slices. I think we may have discussed this and I may have asked you to do it this way, but I may have forgotten the reason why.